include/wally_descriptor.h

```diff
@@ -0,0 +1,172 @@
+#ifndef LIBWALLY_CORE_DESCRIPTOR_H
+#define LIBWALLY_CORE_DESCRIPTOR_H
+
+#include "wally_core.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct wally_map;
+/** An opaque type holding a parsed minscript/descriptor expression */
+struct wally_descriptor;
+
+/* Miniscript type flag */
+#define WALLY_MINISCRIPT_WITNESS_SCRIPT 0x00 /** Witness script */
+#define WALLY_MINISCRIPT_TAPSCRIPT      0x01 /** Tapscript */
+#define WALLY_MINISCRIPT_ONLY           0x02 /** Only allow miniscript (not descriptor) expressions */
+
+#define WALLY_MS_IS_RANGED 0x01 /** Allows key ranges via '*' */
+
+
+/**
+ * Parse an output descriptor or miniscript expression.
+ *
+ * :param descriptor: Output descriptor or miniscript expression to parse.
+ * :param vars_in: Map of variable names to values, or NULL.
+ * :param network: ``WALLY_NETWORK_`` constant descripting the network the
+ *|    descriptor belongs to, or WALLY_NETWORK_NONE for miniscript-only expressions.
+ * :param flags: Include ``WALLY_MINISCRIPT_ONLY`` to disallow descriptor
+ *|    expressions, ``WALLY_MINISCRIPT_TAPSCRIPT`` to use x-only pubkeys, or 0.
+ * :param output: Destination for the resulting parsed descriptor.
+ *|    The descriptor returned should be freed using `wally_descriptor_free`.
+ *
+ * Variable names can be used in the descriptor string and will be substituted
+ * with the contents of ``vars_in`` during parsing. Key names for ``vars_in``
+ * must be 16 characters or less in length and start with a letter.
+ */
+WALLY_CORE_API int wally_descriptor_parse(
```

Comment:
This could use improved error messages / codes. For example I'm getting a `-2` (`WALLY_EINVAL`) when parsing `pkh(xpub68Gmy5EdvgibQVfPdqkBBCHxA5htiqg55crXYuXoQRKfDBFA1WEjWgP6LHhwBZeNK1VTsfTFUHCdrfp1bgwQ9xv5ski8PX9rL2dZXvgGDnw/1)` with network set to `WALLY_NETWORK_BITCOIN_MAINNET`, `flags` set to zero and variables to `NULL`.

Reply:
I've just added this as a test case and it parses without error, please see the commit.

For error handling, I agree both descriptors and PSBT support demand better error messages. The intention here is to support (optional) rich errors for all calls with a `wally_get_error_message` API, but this will be a separate undertaking/PR.

Reply:
I'm not seeing it in `src/test/test_descriptor.py` in 4d430b8182d58387453889dec291450a4ab65cd5. Did you mean to push a new version?

But if it passes, maybe I'm doing something wrong on the Swift - C mapping...

Reply:
Ah yes I haven't pushed this yet, sorry. I will shortly.

Reply:
pushed, as commented above

Reply:
Thanks, it was bug on my end.